File: Task/Hash-join/C-sharp/hash-join.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace HashJoin
{
    public class AgeName
    {
        public AgeName(byte age, string name)
        {
            Age = age;
            Name = name;
        }
        public byte Age { get; private set; }
        public string Name { get; private set; }
    }

    public class NameNemesis
    {
        public NameNemesis(string name, string nemesis)
        {
            Name = name;
            Nemesis = nemesis;
        }
        public string Name { get; private set; }
        public string Nemesis { get; private set; }
    }

    public class DataContext
    {
        public DataContext()
        {
            AgeName = new List<AgeName>();
            NameNemesis = new List<NameNemesis>();
        }
        public List<AgeName> AgeName { get; set; }
        public List<NameNemesis> NameNemesis { get; set; }
    }

    public class AgeNameNemesis
    {
        public AgeNameNemesis(byte age, string name, string nemesis)
        {
            Age = age;
            Name = name;
            Nemesis = nemesis;
        }
        public byte Age { get; private set; }
        public string Name { get; private set; }
        public string Nemesis { get; private set; }
    }

    class Program
    {
        public static void Main()
        {
            var data = GetData();
            var result = ExecuteHashJoin(data);
            WriteResultToConsole(result);
        }

        private static void WriteResultToConsole(List<AgeNameNemesis> result)
        {
            result.ForEach(ageNameNemesis => Console.WriteLine("Age: {0}, Name: {1}, Nemesis: {2}",
                ageNameNemesis.Age, ageNameNemesis.Name, ageNameNemesis.Nemesis));
        }

        private static List<AgeNameNemesis> ExecuteHashJoin(DataContext data)
        {
            return (data.AgeName.Join(data.NameNemesis,
                ageName => ageName.Name, nameNemesis => nameNemesis.Name,
                (ageName, nameNemesis) => new AgeNameNemesis(ageName.Age, ageName.Name, nameNemesis.Nemesis)))
                .ToList();
        }

        private static DataContext GetData()
        {
            var context = new DataContext();

            context.AgeName.AddRange(new [] {
                    new AgeName(27, "Jonah"),
                    new AgeName(18, "Alan"),
                    new AgeName(28, "Glory"),
                    new AgeName(18, "Popeye"),
                    new AgeName(28, "Alan")
                });

            context.NameNemesis.AddRange(new[]
            {
                new NameNemesis("Jonah", "Whales"),
                new NameNemesis("Jonah", "Spiders"),
                new NameNemesis("Alan", "Ghosts"),
                new NameNemesis("Alan", "Zombies"),
                new NameNemesis("Glory", "Buffy")
            });

            return context;
        }
    }
}

File: Task/Hash-join/Zen-C/hash-join.c
import "std/map.zc"
import "std/vec.zc"

struct A {
    age : int;
    name: string;
}

struct B {
    character: string;
    nemesis  : string;
}

fn main() {
    let tableA = [
        A{age: 27, name: "Jonah"}, A{age: 18, name: "Alan"},
        A{age: 28, name: "Glory"}, A{age: 18, name: "Popeye"},
        A{age: 28, name: "Alan"}
    ];

    let tableB = [
        B{character: "Jonah", nemesis: "Whales"},
        B{character: "Jonah", nemesis: "Spiders"},
        B{character: "Alan",  nemesis: "Ghosts"},
        B{character: "Alan",  nemesis: "Zombies"},
        B{character: "Glory", nemesis: "Buffy"}
    ];

    let h = Map<int>::new();
    let vv = Vec<Vec<int>>::new();
    for i in 0..tableA.len {
        let a = tableA[i];
        if h.contains(a.name) {
            let ix = h[a.name].unwrap();
            vv.get_ref(ix).push(i);
        } else {
            let v = Vec<int>::new();
            v.push(i);
            h.put(a.name, (int)vv.length());
            vv.push(v);
        }
    }

    println "Age  Name   Character Nemesis";
    println "---  -----  --------- -------";

    for i in 0..tableB.len {
        let b = tableB[i];
        if h.contains(b.character) {
            let ix = h[b.character].unwrap();
            let f = "%3d  %-5s  %-9s %s\n";
            for e in vv[ix] {
                let t = tableA[e];
                printf(f, t.age, t.name, b.character, b.nemesis);
            }
        }
    }

    for v in vv { v.free(); }
}
